### PR TITLE
Feature/pause resume analytics

### PR DESCRIPTION
**Pull request by achocron**
Added two APIs: `[MSAnalytics pause]` and `[MSAnalytics resume]`.

### CODE REVIEW

File: AppCenter/AppCenterTests/MSOneCollectorChannelDelegateTests.m [hunk 1]
```diff
@@ -101,77 +101,164 @@ - (void)testDidAddChannelUnitWithOneCollectorGroupId {
...
 
-- (void)testDidSetEnabledAndDeleteDataOnDisabledWithBaseGroupId {
-
-  /*
-   * Test base channel unit's logs are cleared when the base channel unit is
-   * disabled. First, add a base channel unit to the channel group. Then,
-   * disable the base channel unit. Lastly, verify the storage deletion is
-   * called for the base channel group id.
-   */
+// TODO Those tests aren't testing anything, they should test that calling setEnable on a base channel unit also disable the corresponding OC channel unit.
```
Comment (achocron) [inline AppCenter/AppCenterTests/MSOneCollectorChannelDelegateTests.m]: Made a bug for this one

**Comment by ElektrojungeMS** [comment]
Shouldn't this go in #1129 or the other way round?

File: AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m [hunk 2]
```diff
@@ -492,20 +492,24 @@ - (void)setEnabled:(BOOL)isEnabled andDeleteDataOnDisabled:(BOOL)deleteData {
...
-- (void)resume {
-  if (self.suspended && self.enabled) {
-    MSLogDebug([MSAppCenter logTag], @"Resume channel for group Id %@.",
-               self.configuration.groupId);
-    self.suspended = NO;
+- (void)resumeWithToken:(id <NSObject>)token {
+  [self.pausedTokens removeObject:token];
+  MSLogDebug([MSAppCenter logTag], @"Pause token %@ removed from channel with group Id %@.",
+             token, self.configuration.groupId);
+  if ([self.pausedTokens count] == 0) {
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m]: It only makes sense if tokens are services like channel, storage and ingestion. What if it is a target token? Not supposed to be used for transmission target pause/resume?

File: AppCenter/AppCenter/Internals/Channel/MSChannelDelegate.h [hunk 3]
```diff
@@ -87,6 +87,22 @@
               didSetEnabled:(BOOL)isEnabled
     andDeleteDataOnDisabled:(BOOL)deletedData;
 
+/**
+ * A callback that is called when pause has been invoked.
+ *
+ * @param channel The channel.
+ * @param token The token used to pause the channel.
+ */
+- (void)channel:(id <MSChannelProtocol>)channel didPauseWithToken:(id <NSObject>)token;
```
Comment (jaelim-ms) [inline AppCenter/AppCenter/Internals/Channel/MSChannelDelegate.h]: We should use different name than `token`.

File: AppCenter/AppCenter/Internals/Channel/MSChannelDelegate.h [hunk 4]
```diff
@@ -87,6 +87,22 @@
               didSetEnabled:(BOOL)isEnabled
     andDeleteDataOnDisabled:(BOOL)deletedData;
 
+/**
+ * A callback that is called when pause has been invoked.
+ *
+ * @param channel The channel.
+ * @param identifyingObject The identifying object used to pause the channel.
```
Comment (ElektrojungeMS) [inline AppCenter/AppCenter/Internals/Channel/MSChannelDelegate.h]: I like the API but I'd love to see some more explanation around this, e.g. what object the `identifyingObject` could be, and what "pausing" means. It's not a release blocker, just my opinion.

File: AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m [hunk 5]
```diff
@@ -498,22 +498,48 @@ - (void)setEnabled:(BOOL)isEnabled andDeleteDataOnDisabled:(BOOL)deleteData {
...
 }
 
-- (void)suspend {
-  if (!self.suspended) {
-    MSLogDebug([MSAppCenter logTag], @"Suspend channel for group Id %@.",
-               self.configuration.groupId);
-    self.suspended = YES;
+- (void)pauseWithIdentifyingObject:(id <NSObject>)identifyingObject {
+  dispatch_async(self.logsDispatchQueue, ^{
+    [self pauseWithIdentifyingObjectSync:identifyingObject];
```
Comment (ElektrojungeMS) [inline AppCenter/AppCenter/Internals/Channel/MSChannelUnitDefault.m]: I like that we have `sync` in the name here.

File: AppCenter/AppCenterTests/MSChannelGroupDefaultTests.m [hunk 6]
```diff
@@ -11,10 +11,27 @@
...
+  MSPriority priority = MSPriorityDefault;
+  float flushInterval = 1.0;
+  NSUInteger batchSizeLimit = 10;
+  NSUInteger pendingBatchesLimit = 3;
+  _validConfiguration = [[MSChannelUnitConfiguration alloc]
+                       initWithGroupId:groupId
+                       priority:priority
+                       flushInterval:flushInterval
+                       batchSizeLimit:batchSizeLimit
+                         pendingBatchesLimit:pendingBatchesLimit];
```
Comment (jaelim-ms) [inline AppCenter/AppCenterTests/MSChannelGroupDefaultTests.m]: Weird format but please do not change.

File: AppCenterAnalytics/AppCenterAnalyticsTests/MSAnalyticsTests.m [hunk 7]
```diff
@@ -969,4 +969,52 @@ - (void)testDictionaryContainsValidNestedProperties {
...
+  
+  // When
+  [MSAnalytics pause];
+  
+  // Then
+  OCMVerify([[MSAnalytics sharedInstance].channelUnit pauseWithIdentifyingObject:[MSAnalytics sharedInstance]]);
+  [appCenterMock stopMocking];
+}
+
+-(void)testResume {
```
[the remark this answers is not in the document]
Comment (clpolet) [reply AppCenterAnalytics/AppCenterAnalyticsTests/MSAnalyticsTests.m]: Done in channel tests :)

File: AppCenterAnalytics/AppCenterAnalytics/MSAnalytics.h [hunk 8]
```diff
@@ -24,6 +24,21 @@ NS_ASSUME_NONNULL_BEGIN
 + (void)trackEvent:(NSString *)eventName
     withProperties:(nullable NSDictionary<NSString *, NSString *> *)properties;
 
+/**
+ * Pause transmission of Analytics logs. While paused, Analytics logs are saved to disk.
```
[the remark this answers is not in the document]
Comment (clpolet) [reply AppCenterAnalytics/AppCenterAnalytics/MSAnalytics.h]: It feels fine to me since it's specified for Analytics logs.

**Pull request merged**